### ISSUE TITLE
Heap out of bounds read in lookup_bytestring() 2

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#553]
**Heap out of bounds read in lookup_bytestring() 2**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_lookup_bytestring_2.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20090811/af7229e2-a58f-11e6-95c9-47af070c5dc5.png)

ASAN output:

```
==15069==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000ef3c at pc 0x0000004a13b7 bp 0x7ffd79d551b0 sp 0x7ffd79d54960
READ of size 8 at 0x60300000ef3c thread T0
    #0 0x4a13b6 in __asan_memcpy /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_interceptors.cc:413:3
    #1 0x4f2f89 in lookup_bytestring XYZ/tcpdump/./addrtoname.c:398:2
    #2 0x4f32fd in linkaddr_string XYZ/tcpdump/./addrtoname.c:570:7
    #3 0x576f02 in print_long_pos_vector XYZ/tcpdump/./print-geonet.c:89:2
    #4 0x576f02 in geonet_print XYZ/tcpdump/./print-geonet.c:172
    #5 0x566e7a in ethertype_print XYZ/tcpdump/./print-ether.c:418:17
    #6 0x5663f8 in ether_print XYZ/tcpdump/./print-ether.c:229:7
    #7 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #8 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #9 0x7ae607 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #10 0x6b7a7c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #11 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #12 0x7f815f5c582f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #13 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x60300000ef3c is located 0 bytes to the right of 28-byte region [0x60300000ef20,0x60300000ef3c)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7afc13 in pcap_check_header XYZ/libpcap/./sf-pcap.c:401:14
    #2 0x7ada32 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7ad7bf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7f815f5c582f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_interceptors.cc:413:3 in __asan_memcpy
Shadow bytes around the buggy address:
  0x0c067fff9d90: 00 00 00 00 fa fa 00 00 00 01 fa fa 00 00 04 fa
  0x0c067fff9da0: fa fa 00 00 07 fa fa fa 00 00 00 01 fa fa 00 00
  0x0c067fff9db0: 00 02 fa fa 00 00 00 01 fa fa 00 00 06 fa fa fa
  0x0c067fff9dc0: 00 00 00 fa fa fa 00 00 07 fa fa fa 00 00 06 fa
  0x0c067fff9dd0: fa fa 00 00 07 fa fa fa 00 00 00 00 fa fa fd fd
=>0x0c067fff9de0: fd fa fa fa 00 00 00[04]fa fa 00 00 04 fa fa fa
  0x0c067fff9df0: 00 00 05 fa fa fa 00 00 04 fa fa fa 00 00 04 fa
  0x0c067fff9e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9e30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==15069==ABORTING
```

**Comment by fxlb** [comment]
Thanks for the report.
This issue is already fixed in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit 6bc4429 fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.
